### PR TITLE
[tune] Handle infinite and NaN values

### LINKED ISSUE #11670: [tune] Schedulers should automatically handle nans/inf values
### What is the problem?

I get the following when I run the following code
```
  def tune_asha(num_samples=400, num_epochs=25, gpus_per_trial=0.1):
        config = {
            "logger_config": {"mlflow_experiment_id": experiment_id},
            "batch_size": 1024,
            "learning_rate": tune.uniform(1e-5, 1e-3),
            "layer1_2_size": tune.uniform(128, 512),
            "layer3": tune.uniform(64, 256),
            "layer4": tune.uniform(32, 128),
            "layer5": tune.uniform(16, 64),
            "layer1_2_d": tune.uniform(0.01, 0.2),
            "layer3_4_5_d": tune.uniform(0.01, 0.05),
            "emb_dropout": tune.uniform(0.01, 0.1),
            "emb_max": tune.uniform(64, 256),
            "gradient_clipping": tune.uniform(5, 500),
            "accul_batches": 10,
            "optimizer_weight_decay": tune.uniform(0, 0.001),
            "lr_scheduler_patience": tune.uniform(1, 5),
            "lr_scheduler_reduction_factor": tune.uniform(0.1, 0.25),
        }

        scheduler = ASHAScheduler(
            metric="loss",
            mode="min",
            max_t=num_epochs,
            grace_period=7,
            reduction_factor=2,
        )

        reporter = CLIReporter(
            parameter_columns=[
                "layer1_2_size",
                "layer3",
                "layer4",
                "layer5",
                "learning_rate",
                "batch_size",
            ],
            metric_columns=["loss", "training_iteration"],
        )

        algo = BayesOptSearch(
            metric="loss",
            mode="min",
            utility_kwargs={"kind": "ucb", "kappa": 2.5, "xi": 0.0},
            random_search_steps=30,
        )
        algo = ConcurrencyLimiter(algo, max_concurrent=10)

        return tune.run(
            partial(train_tune, num_epochs=num_epochs, num_gpus=gpus_per_trial),
            name=experiment_name,
            resources_per_trial={"cpu": 1, "gpu": gpus_per_trial},
            config=config,
            local_dir=".",
            search_alg=algo,
            scheduler=scheduler,
            num_samples=num_samples,
            progress_reporter=reporter,
            loggers=DEFAULT_LOGGERS + (MLFLowLogger,),
        )
```

When it's seeding the space there are no issues but as soon as it is on trial 30, I get the following error

```Traceback (most recent call last):
  File "tune_r_learner_bo.py", line 116, in <module>
    main()
  File "tune_r_learner_bo.py", line 110, in main
    expr_obj = tune_asha()
  File "tune_r_learner_bo.py", line 107, in tune_asha
    loggers=DEFAULT_LOGGERS + (MLFLowLogger,),
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/tune.py", line 411, in run
    runner.step()
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 561, in step
    next_trial = self._get_next_trial()  # blocking
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 658, in _get_next_trial
    self._update_trial_queue(blocking=wait_for_trial)
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 1011, in _update_trial_queue
    trial = self._search_alg.next_trial()
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/suggest/search_generator.py", line 114, in next_trial
    self._experiment.dir_name)
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/suggest/search_generator.py", line 121, in create_trial_if_possible
    suggested_config = self.searcher.suggest(trial_id)
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/suggest/suggestion.py", line 346, in suggest
    suggestion = self.searcher.suggest(trial_id)
  File "/opt/conda/lib/python3.7/site-packages/ray/tune/suggest/bayesopt.py", line 257, in suggest
    config = self.optimizer.suggest(self.utility)
  File "/opt/conda/lib/python3.7/site-packages/bayes_opt/bayesian_optimization.py", line 128, in suggest
    self._gp.fit(self._space.params, self._space.target)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/gaussian_process/_gpr.py", line 190, in fit
    ensure_2d=True, dtype="numeric")
  File "/opt/conda/lib/python3.7/site-packages/sklearn/base.py", line 432, in _validate_data
    X, y = check_X_y(X, y, **check_params)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 805, in check_X_y
    ensure_2d=False, dtype=None)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 72, in inner_f
    return f(**kwargs)
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 645, in check_array
    allow_nan=force_all_finite == 'allow-nan')
  File "/opt/conda/lib/python3.7/site-packages/sklearn/utils/validation.py", line 99, in _assert_all_finite
    msg_dtype if msg_dtype is not None else X.dtype)
ValueError: Input contains NaN, infinity or a value too large for dtype('float64').
```

My neural network is capable of inf and Nan losses but will not always generate these values. Further I would expect there to be a workaround if it does produce a Nan I just haven't been able to find it.

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

I'm in a docker container that is based off the latest pytorch image. 

I'm building Ray from the latest wheel (or the latest from Monday) 
https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-1.1.0.dev0-cp37-cp37m-manylinux1_x86_64.whl
ray==1.1.0.dev0
Python 3.7.7
bayesian-optimization==1.2.0

If we cannot run your script, we cannot fix your issue.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

If NaN or infinite values are reported, some searchers crash. This PR introduces a utility function to handle these cases. Additionally, trials reporting infinite values will be excluded from the `analysis.get_best_trial()` method per default.

This is a draft PR with support for searchers. Schedulers will be handled next.

## Related issue number

Closes #11670 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/tests/test_searchers.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        tune.report(float(config[metric]) or 0.1)
+
+
+class InvalidValuesTest(unittest.TestCase):
+    def setUp(self):
+        self.config = {"report": tune.uniform(0.0, 5.0)}
+
+    def tearDown(self):
+        pass
+
```
Comment (richardliaw) [inline python/ray/tune/tests/test_searchers.py:28]: to speed this up, can you do:

```
    @classmethod
    def setUpClass(cls):
        ray.init(
            num_cpus=4, num_gpus=0, local_mode=True, include_dashboard=False)

    @classmethod
    def tearDownClass(cls):
        ray.shutdown()
        _register_all()
```

maybe you'll need to remove the `reuse_actors`.

File: python/ray/tune/tests/test_searchers.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            _invalid_objective,
+            search_alg=ZOOptSearch(budget=8, parallel_num=8),
+            config=self.config,
+            metric="_metric",
+            mode="max",
+            num_samples=8,
+            reuse_actors=True)
+
+        best_trial = out.best_trial
+        self.assertLessEqual(best_trial.config["report"], 2.0)
```
Comment (richardliaw) [inline python/ray/tune/tests/test_searchers.py:149]: does it make sense to abstract this into a helper function?

**Review by richardliaw — COMMENTED** [review]
Looks good! Can you somehow document this behavior in the docstring of the objects?

File: python/ray/tune/tests/test_searchers.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/tests/test_searchers.py:149]: There's some special cases for some of the algorithms for seeding in the last changes, thus I'd like to keep it as is for now

**Review by richardliaw — APPROVED** [review]

**Pull request merged**